workitem/workitem_repository.go

```diff
@@ -1131,3 +1161,191 @@ func (r *GormWorkItemRepository) LoadByIteration(ctx context.Context, iterationI
 	}
 	return workitems, nil
 }
+
+// ChangeWorkItemType changes the workitem in wiStorage to newWIType. Returns error if the operation fails
+func (r *GormWorkItemRepository) ChangeWorkItemType(ctx context.Context, wiStorage *WorkItemStorage, oldWIType *WorkItemType, newWIType *WorkItemType, spaceID uuid.UUID) error {
+	allowedWIT, err := r.CheckWIT(ctx, newWIType, spaceID)
+	if err != nil {
+		return errs.Wrap(err, "failed to check workitem type")
+	}
+	if !allowedWIT {
+		return errors.NewBadParameterError("typeID", oldWIType.ID)
+	}
+	var fieldDiff = Fields{}
+	// Loop through old workitem type
+	for oldFieldName := range oldWIType.Fields {
+		newFieldType := newWIType.Fields[oldFieldName].Type
+		// Temporary workaround to not add metastates to the field diff
+		// We need to have a special handling for fields are shouldn't be set by user (or affected by type change)
+		if oldFieldName == SystemMetaState {
+			continue
+		}
+		// The field exists in old type and new type
+		if _, ok := newWIType.Fields[oldFieldName]; ok {
+			// Try to assign the old value to the new field
+			_, err := newFieldType.ConvertToModel(wiStorage.Fields[oldFieldName])
+			if err != nil {
+				// The field might be a list type. Try to assign it to a list
+				if newFieldType.GetKind() == KindList {
+					interfaceArray, ok := wiStorage.Fields[oldFieldName].([]interface{})
+					if ok {
+						_, err = newFieldType.ConvertToModel(interfaceArray)
+					}
+				}
+			}
+			if err != nil {
+				// Failed to assign the new value to the old field
+				fieldDiff[oldFieldName] = wiStorage.Fields[oldFieldName]
+				delete(wiStorage.Fields, oldFieldName)
+			}
+		} else { // field doesn't exist in new type
+			if wiStorage.Fields[oldFieldName] != nil {
+				fieldDiff[oldFieldName] = wiStorage.Fields[oldFieldName]
+				delete(wiStorage.Fields, oldFieldName)
+			}
+		}
+	}
+	// We need fieldKeys to show field diff in a defined order. Golang maps aren't ordered by default.
+	var fieldKeys []string
+	for fieldName := range fieldDiff {
+		fieldKeys = append(fieldKeys, fieldName)
+	}
+	// Sort the field keys to prevent random order of fields
+	sort.Strings(fieldKeys)
+	// Append diff (fields along with their values) between the workitem types to the description
+	if len(fieldDiff) > 0 {
+		originalDescription := rendering.NewMarkupContentFromValue(wiStorage.Fields[SystemDescription])
+		textToPrepend := "\n\n======= Missing fields in workitem type: " + newWIType.Name + " =======\n"
```

Comment:
@kwk I feel it would be an overkill. We'll have to create 
1. A template.
2. A data structure that holds the actual data.
3. Parse the template and add all data to it (maybe just 2 function calls).

Do you still think we should do it? I started implementing and then realized that this might be overkill.

Reply:
Thanks for the help @kwk . I've used text/template package to render description. Please see https://github.com/fabric8-services/fabric8-wit/pull/2202/commits/5ee62b25ac0b6851eb25ff39a605e2c205a875b9